src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculator.java

```diff
@@ -0,0 +1,167 @@
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.enums.AccrualType;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.timecard.TimeEntry;
+
+@Component
+public class BalanceCalculator {
+
+  private final RestClient restClient;
+
+  @Autowired
+  public BalanceCalculator(RestClient restClient) {
+    this.restClient = restClient;
+  }
+
+  public List<Accrual> calculate(TimeEntry timeEntry) {
+
+    List<Accrual> accrualsToBatchUpdate = Arrays.stream(AccrualType.values())
+        .filter(AccrualType::isEnabled)
+        .map(accrualType -> calculateAccruals(timeEntry, accrualType))
+        .flatMap(Collection::stream)
+        .toList();
+
+    // TODO : send Batch Update request to Accruals API
+    // restClient.batchUpdate(accrualsToBatchUpdate)
+
+    return accrualsToBatchUpdate;
+  }
+
+  List<Accrual> calculateAccruals(TimeEntry timeEntry, AccrualType accrualType) {
+
+    Map<LocalDate, Range<ZonedDateTime>> dateRangeMap =
+        splitOverDays(timeEntry.getActualStartTime(), timeEntry.getActualEndTime());
+
+    return dateRangeMap.entrySet().stream()
+        .map(entry -> recalculateContributions(timeEntry.getId(), timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrualType, entry.getKey(), entry.getValue()))
+        .map(accrual -> cascadeCumulativeTotal(timeEntry, accrual)).flatMap(Collection::stream)
+        .toList();
+  }
+
+  private List<Accrual> cascadeCumulativeTotal(TimeEntry timeEntry, Accrual accrual) {
+    LocalDate referenceDate = accrual.getAccrualDate();
+
+    List<Accrual> accrualsToUpdate =
+        getAccrualsFromReferenceDateUntilEndOfAgreement(timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrual.getAgreementId().toString(), referenceDate.plusDays(1));
+    accrualsToUpdate.add(0, accrual);
+
+    BigDecimal priorCumulativeTotal =
+        restClient.getPriorAccrual(timeEntry.getTenantId(), timeEntry.getOwnerId(),
+            accrual.getAccrualTypeId().toString(), referenceDate).getCumulativeTotal();
+
+    return updateSubsequentAccruals(accrualsToUpdate, priorCumulativeTotal);
+  }
+
+  List<Accrual> updateSubsequentAccruals(List<Accrual> accruals, BigDecimal priorCumulativeTotal) {
+
+    List<Accrual> updatedAccruals = List.copyOf(accruals);
+    //update the cumulative total for referenceDate
+    updatedAccruals.get(0).setCumulativeTotal(
+        priorCumulativeTotal.add(updatedAccruals.get(0).getContributions().getTotal()));
+
+    //cascade through until end of agreement
+    for (int i = 1; i < updatedAccruals.size(); i++) {
+      BigDecimal priorTotal =
+          updatedAccruals.get(i - 1).getCumulativeTotal();
+      Accrual currentAccrual = updatedAccruals.get(i);
+      currentAccrual.setCumulativeTotal(
+          priorTotal.add(currentAccrual.getContributions().getTotal()));
+    }
+
+    return updatedAccruals;
+  }
+
+  Accrual recalculateContributions(String timeEntryId, String tenantId,
```

Comment:
Feels like this is specific to standard hours. Should there be an abstraction i.e. this class is for ensuring the modules calculate the contributions but the actual contribution should be calculated by each module

Reply:
This logic moved to new `AccrualModule` abstract class. New method is called `updateAccrualContribution()`. Child modules can override this method if needed

Reply:
Looks like this has evolved a little further than your last comment. So the accrual module has the responsibility of calculating the contribution `calculateShiftContribution` but the calculator has the responsibility of applying the contribution to the accruals in the method `updateAccrualContribution`. This looks good to me, I just wanted to clarify I understand the current logic correctly and if so I'm happy for you to resolve this conversation

Reply:
The most recent implementation is indeed inline with your understanding